### PR TITLE
Replicated structure of model buyer

### PR DESCRIPTION
**Pull request by GFibrizo**
This PR addresses the issue #2 by refactoring the code to:
- use flask-restplus for the API
- use sqlalchemy as client to manage the persistence into a SQLite3 database
- have a login with user session

This PR is compatible with the changes introduced by the federated-aggregator's PR [6](https://github.com/DeltaML/federated-aggregator/pull/6)
This PR is compatible with the changes introduced by the federated-aggregator's PR [7](https://github.com/DeltaML/federated-aggregator/pull/7)

### CODE REVIEW

File: commons/data/data_loader.py [hunk 1]
```diff
@@ -104,6 +110,8 @@ def get_dataset_for_training(self, requeriments):
         :return: the name of the file that complies with the requested requirements.
         """
         files = [fname for fname in os.listdir(self.dataset_path)]
+        print(requeriments['features'])
```
Comment (agrojas) [inline commons/data/data_loader.py]: Sacar prints

File: data_owner/config/prod/app_config.py [hunk 2]
```diff
@@ -4,4 +4,5 @@
 REGISTRATION_ENABLE = True
 ACTIVE_ENCRYPTION = False
 DATASETS_DIR = "./dataset"
-KEY_LENGTH = 1024
\ No newline at end of file
+KEY_LENGTH = 1024
+DB_ENGINE = 'sqlite:////app/db/model_buyer.db'
```
Comment (agrojas) [inline data_owner/config/prod/app_config.py]: Debería tener su propia db

File: data_owner/models/training.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class Status(Enum):
+    WAITING = 0
+    INITIATED = 1
+    IN_PROGRESS = 2
+    FINISHED = 3
+
+
+class Training(DbEntity):
```
Comment (agrojas) [inline data_owner/models/training.py]: Borrar

File: data_owner/resources/__init__.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+# Add apis to namespace
+api.add_namespace(trainings_api)
+api.add_namespace(datasets_api)
+api.add_namespace(metrics_api)
+
+
+@api.errorhandler(LoginFailureException)
+def login_failure_handler(error):
```
Comment (agrojas) [inline data_owner/resources/__init__.py]: utilizar codigo de error que viene de la excepcion

File: data_owner/resources/metrics_resource.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    'mse': fields.Float(required=True, description='The model mse')
+})
+
+data_owner = DataOwnerService()
+
+
+@api.route('', endpoint='metrics_resources_ep')
+class MetricsResources(Resource):
+
+    @api.doc('Create new user')
```
Comment (agrojas) [inline data_owner/resources/metrics_resource.py]: Cambiar descripcion

File: data_owner/resources/trainings_resource.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    'has_dataset': fields.Boolean(required=True, description='The model weights')
+})
+
+data_owner = DataOwnerService()
+
+
+@api.route('', endpoint='training_resources_ep')
+class TrainingResources(Resource):
+
+    @api.doc('Create new user')
```
Comment (agrojas) [inline data_owner/resources/trainings_resource.py]: Descripcion

File: data_owner/resources/trainings_resource.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        """
+        data = request.get_json()
+        logging.info('Gradient step')
+        model_orm = Model.get(model_id)
+        data_owner.step(model_orm.model, data['gradient'])
+        model_orm.update()
+        return 200
+
+
+#@api.route('/<model_id>', endpoint='training_ep')
```
Comment (agrojas) [inline data_owner/resources/trainings_resource.py]: Voletear

File: data_owner/resources/trainings_resource.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    @api.doc('Create new user')
+    @api.marshal_with(link, code=201)
+    def post(self):
+        data = request.get_json()
+        training_id = data['model_id']
+        reqs = data['requirements']
+        # TODO: For now i'm creating the training and linking the dataset to the training all at once
+        # TODO: and doing it in the back, but a future change will be to do those in separate API calls.
+        filename = DataLoader().get_dataset_for_training(reqs)
+        if filename is not None:
```
Comment (agrojas) [inline data_owner/resources/trainings_resource.py]: Pasar a un service

File: data_owner/resources/trainings_resource.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    @api.doc('Get gradient updated')
+    @api.marshal_with(update, code=200)
+    def post(self, model_id):
+        """
+        Process weights from server
+        :return:
+        """
+        logging.info('Process weights')
+        data = request.get_json()
+        model_orm = Model.get(model_id)
```
Comment (agrojas) [inline data_owner/resources/trainings_resource.py]: Pasar a un service y quitar print

File: commons/data/data_loader.py (hunk 1 above)
Comment (GFibrizo) [reply commons/data/data_loader.py]: Done

File: data_owner/config/prod/app_config.py (hunk 2 above)
Comment (GFibrizo) [reply data_owner/config/prod/app_config.py]: Done.

File: data_owner/models/training.py (hunk 3 above)
Comment (GFibrizo) [reply data_owner/models/training.py]: Done

File: data_owner/resources/metrics_resource.py (hunk 5 above)
Comment (GFibrizo) [reply data_owner/resources/metrics_resource.py]: Done

File: data_owner/resources/trainings_resource.py (hunk 6 above)
Comment (GFibrizo) [reply data_owner/resources/trainings_resource.py]: Done

File: data_owner/resources/trainings_resource.py (hunk 7 above)
Comment (GFibrizo) [reply data_owner/resources/trainings_resource.py]: Done

File: data_owner/resources/trainings_resource.py (hunk 8 above)
Comment (GFibrizo) [reply data_owner/resources/trainings_resource.py]: Done

File: data_owner/resources/trainings_resource.py (hunk 9 above)
Comment (GFibrizo) [reply data_owner/resources/trainings_resource.py]: Done

File: data_owner/services/data_owner_service.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
+import logging
+import uuid
+
+import numpy as np
+
+from commons.decorators.decorators import optimized_collection_parameter
+from commons.utils.singleton import Singleton
+from commons.data.data_loader import DataLoader
+from commons.model.model_service import ModelFactory
+from data_owner.domain.data_owner import DataOwner
```
Comment (agrojas) [inline data_owner/services/data_owner_service.py]: Falta subir esto

**Pull request merged**